File: experiments/puppeteer-automation/saucedemo-e2e/tests/inventory.test.ts
import assert from 'node:assert/strict';
import { describe, it } from 'node:test';
import { InventoryPage } from '../src/pages/saucedemo.pages.js';
import { PRODUCTS } from '../src/utils/test-data.js';
import { withAuthenticatedPage } from './harness.js';

describe('Inventory', () => {
  it('displays product catalog', async () => {
    await withAuthenticatedPage('inventory-catalog', async (page) => {
      const inventory = new InventoryPage(page);
      assert.equal(await inventory.getProductCount(), 6);
    });
  });

  it('adds product to cart and updates badge', async () => {
    await withAuthenticatedPage('inventory-add', async (page) => {
      const inventory = new InventoryPage(page);
      await inventory.addToCart(PRODUCTS.backpack);
      assert.equal(await inventory.getCartBadgeText(), '1');
    });
  });

  it('removes product from cart on inventory page', async () => {
    await withAuthenticatedPage('inventory-remove', async (page) => {
      const inventory = new InventoryPage(page);
      await inventory.addToCart(PRODUCTS.bikeLight);
      await inventory.removeFromCart(PRODUCTS.bikeLight);
      assert.equal(await inventory.hasCartBadge(), false);
    });
  });

  it('sorts products A to Z', async () => {
    await withAuthenticatedPage('inventory-sort-az', async (page) => {
      const inventory = new InventoryPage(page);
      await inventory.sortBy('az');
      const names = await inventory.getProductNames();
      const sorted = [...names].sort((a, b) => a.localeCompare(b));
      assert.deepEqual(names, sorted);
    });
  });

  it('sorts products Z to A', async () => {
    await withAuthenticatedPage('inventory-sort-za', async (page) => {
      const inventory = new InventoryPage(page);
      await inventory.sortBy('za');
      const names = await inventory.getProductNames();
      const sorted = [...names].sort((a, b) => b.localeCompare(a));
      assert.deepEqual(names, sorted);
    });
  });

  it('logs out to login page', async () => {
    await withAuthenticatedPage('inventory-logout', async (page) => {
      const inventory = new InventoryPage(page);
      await inventory.logout();
      assert.match(page.url(), /saucedemo\.com\/?$/);
    });
  });
});

File: experiments/selenium-java/saucedemo-e2e/src/test/java/com/qaskills/saucedemo/tests/InventoryTest.java
package com.qaskills.saucedemo.tests;

import com.qaskills.saucedemo.pages.InventoryPage;
import com.qaskills.saucedemo.pages.LoginPage;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

public class InventoryTest extends BaseTest {

    private static final String BACKPACK = "Sauce Labs Backpack";
    private static final String BIKE_LIGHT = "Sauce Labs Bike Light";

    private InventoryPage inventoryPage;

    @BeforeMethod(dependsOnMethods = "setUp")
    public void login() {
        inventoryPage = loginAsStandardUser();
    }

    @Test(description = "Verify product catalog is displayed")
    public void testProductCatalog() {
        assertThat(inventoryPage.getProductCount()).isEqualTo(6);
    }

    @Test(description = "Verify add to cart updates badge")
    public void testAddProductToCart() {
        inventoryPage.addProductToCart(BACKPACK);

        assertThat(inventoryPage.getCartBadgeText()).isEqualTo("1");
        assertThat(inventoryPage.isAddToCartVisible(BACKPACK)).isFalse();
        assertThat(inventoryPage.isRemoveVisible(BACKPACK)).isTrue();
    }

    @Test(description = "Verify remove product from inventory page")
    public void testRemoveProductFromCart() {
        inventoryPage.addProductToCart(BIKE_LIGHT);
        inventoryPage.removeProductFromCart(BIKE_LIGHT);

        assertThat(inventoryPage.isCartBadgeDisplayed()).isFalse();
    }

    @Test(description = "Verify sort products A to Z")
    public void testSortProductsAscending() {
        inventoryPage.sortBy("az");
        List<String> names = inventoryPage.getProductNames();
        List<String> sorted = new ArrayList<>(names);
        sorted.sort(Comparator.naturalOrder());

        assertThat(names).isEqualTo(sorted);
    }

    @Test(description = "Verify sort products Z to A")
    public void testSortProductsDescending() {
        inventoryPage.sortBy("za");
        List<String> names = inventoryPage.getProductNames();
        List<String> sorted = new ArrayList<>(names);
        sorted.sort(Comparator.reverseOrder());

        assertThat(names).isEqualTo(sorted);
    }

    @Test(description = "Verify logout returns to login page")
    public void testLogout() {
        LoginPage loginPage = inventoryPage.logout();

        assertThat(loginPage.isLoginButtonDisplayed()).isTrue();
        assertThat(loginPage.getCurrentUrl()).doesNotContain("inventory.html");
    }
}
